[WIP] Add Grover Optimization to Optimization Stack (#847)

AustinGilliam opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary

Create a Grover Optimization method for finding the minimum of a QUBO problem, as discussed in https://arxiv.org/abs/1912.04088.

### Details and comments

- Please review "evaluate_classically" for the CustomCircuitOracle. It was unclear to me what this should return, as I was having trouble finding documentation in the other oracles that use this method.
- Is it possible to substitute LogicalExpressionOracle for CustomCircuitOracle?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/qiskit-aqua?pullRequest=848) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/qiskit-aqua?pullRequest=848) before we can accept your contribution.<br/><hr/>**Austin Gilliam** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/qiskit-aqua?pullRequest=848) it.</sub>

AustinGilliam commented:
@stefan-woerner I've gone through and updated the code as per the contribution guidelines, coding style, and pylint recommendations. As for next steps:

1. See above for my initial questions about the type of oracle we use, and evaluate_classically().
2. I had to scrub my email from my first commit (for reasons we discussed previously), and added my personal later. However, now the CLA helper is looking for a signature from a user that doesn't exist. Is there a way to fix/bypass this?
3. Not sure why the build is failing in the doc/style/lint step, all I see in the log is: "Makefile:56: recipe for target 'spell' failed". Do you know the fix for this?

woodsp-ibm commented:
Hi Austin, 
2. CLA. As you have seen the earlier commits with no email are causing an issue. You may be able to do some interactive rebase to have these combined into a commit with your email. Otherwise my only other suggestion is  copy your changed files, re-fork the repo, and apply these files again in new commit on your later id with the email.
3. We have a .pylintdict file in the root to which names etc that are not in the standard spelling dict can be added. So if the spelling is actually correct (maybe iter in your case could be changed to iterations) then add an entry to the .pylintdict custom dict - we keep it in alphabetic order to manage it easier.

manoelmarques commented:
Just tried an interactive rebase locally on this fork to amend the old commits with right author and I don't think this is a safe solution. The updated commit is reapplied but in order for it to happen, it is reapplying posterior commits that are already pushed in Aqua master and creating a brand new commit hash that is different from the one already in master.

AustinGilliam commented:
@manoelmarques @woodsp-ibm Thanks, looks like the CLA helper is happy now. I'll fix the spelling issue shortly.

AustinGilliam commented:
Working with @stefan-woerner to consolidate optimization changes on another fork. Closing this PR and moving the code there.
